test/opt/loop_optimizations/unroll_simple.cpp

```diff
@@ -0,0 +1,2228 @@
+%150 = OpLabel
+%151 = OpIAdd %11 %140 %21
+OpBranch %152
+%152 = OpLabel
+OpBranch %153
+%153 = OpLabel
+%154 = OpIAdd %6 %25 %21
+OpBranch %27
+%27 = OpLabel
+OpReturn
+OpFunctionEnd
+)";
+  // clang-format on
+  std::unique_ptr<ir::IRContext> context =
+      BuildModule(SPV_ENV_UNIVERSAL_1_1, nullptr, text,
+                  SPV_TEXT_TO_BINARY_OPTION_PRESERVE_NUMERIC_IDS);
+  ir::Module* module = context->module();
+  EXPECT_NE(nullptr, module) << "Assembling failed for ushader:\n"
+                             << text << std::endl;
+  opt::LoopUnroller loop_unroller;
+  SetDisassembleOptions(SPV_BINARY_TO_TEXT_OPTION_NO_HEADER);
+  SinglePassRunAndCheck<opt::LoopUnroller>(text, output, false);
+}
+
+/*
+Generated from the following GLSL
+#version 410 core
+void main() {
+    for (int i = 0; i < 10; i++) {
+      if (i == 5) break;
+    }
+}
+*/
+TEST_F(PassClassTest, UnrollWithBreak) {
+  // clang-format off
+  // With opt::LocalMultiStoreElimPass
+const std::string text = R"(
+               OpCapability Shader
+          %1 = OpExtInstImport "GLSL.std.450"
+               OpMemoryModel Logical GLSL450
+               OpEntryPoint Fragment %4 "main"
+               OpExecutionMode %4 OriginUpperLeft
+               OpSource GLSL 410
+               OpName %4 "main"
+          %2 = OpTypeVoid
+          %3 = OpTypeFunction %2
+          %6 = OpTypeInt 32 1
+          %7 = OpTypePointer Function %6
+          %9 = OpConstant %6 0
+         %16 = OpConstant %6 10
+         %17 = OpTypeBool
+         %20 = OpConstant %6 5
+         %26 = OpConstant %6 1
+          %4 = OpFunction %2 None %3
+          %5 = OpLabel
+               OpBranch %10
+         %10 = OpLabel
+         %28 = OpPhi %6 %9 %5 %27 %13
+               OpLoopMerge %12 %13 Unroll
+               OpBranch %14
+         %14 = OpLabel
+         %18 = OpSLessThan %17 %28 %16
+               OpBranchConditional %18 %11 %12
+         %11 = OpLabel
+         %21 = OpIEqual %17 %28 %20
+               OpSelectionMerge %23 None
+               OpBranchConditional %21 %22 %23
+         %22 = OpLabel
+               OpBranch %12
+         %23 = OpLabel
+               OpBranch %13
+         %13 = OpLabel
+         %27 = OpIAdd %6 %28 %26
+               OpBranch %10
+         %12 = OpLabel
+               OpReturn
+               OpFunctionEnd
+)";
+
+const std::string expected = R"(OpCapability Shader
+%1 = OpExtInstImport "GLSL.std.450"
+OpMemoryModel Logical GLSL450
+OpEntryPoint Fragment %2 "main"
+OpExecutionMode %2 OriginUpperLeft
+OpSource GLSL 410
+OpName %2 "main"
+%3 = OpTypeVoid
+%4 = OpTypeFunction %3
+%5 = OpTypeInt 32 1
+%6 = OpTypePointer Function %5
+%7 = OpConstant %5 0
+%8 = OpConstant %5 10
+%9 = OpTypeBool
+%10 = OpConstant %5 5
+%11 = OpConstant %5 1
+%2 = OpFunction %3 None %4
+%12 = OpLabel
+OpBranch %13
+%13 = OpLabel
+OpBranch %18
+%18 = OpLabel
+%19 = OpSLessThan %9 %7 %8
+OpBranch %20
+%20 = OpLabel
+%21 = OpIEqual %9 %7 %10
+OpSelectionMerge %22 None
```

Comment:
This isn't correct. I know it passes the validator. I've filed #1281 to fix that. I would suggest not handling breaks for now at least. Add a rule to CanPerformUnroll to check that the loop has only one exit block. Continues are also potentially troublesome. Structured control flow rules do not require an OpSelectionMerge in the case of breaks and continues. I would suggest adding a check in CanPerformUnroll to verify that any conditional branches to the continue block are preceeded by an OpSelectionMerge. If they are not, I would refuse to unroll them for now.